### PR TITLE
Use typepaths for the quirk blacklist

### PR DESCRIPTION
**Pull request by spookydonut**
## About The Pull Request
The string list is awful for maintainability and is the reason the filter_invalid_quirks() proc wouldn't work properly.

randomise_quirks() is still broken though.

## Changelog
:cl:
fix: Incompatible quirks in existing savefiles shouldn't be possible anymore.
/:cl:

### CODE REVIEW

File: code/controllers/subsystem/processing/quirks.dm [hunk 1]
```diff
@@ -18,29 +18,37 @@ PROCESSING_SUBSYSTEM_DEF(quirks)
...
+	var/static/list/quirk_string_blacklist = list()
+
 /datum/controller/subsystem/processing/quirks/Initialize()
 	get_quirks()
+	quirk_string_blacklist.Cut()
+	for(var/blacklist in quirk_blacklist)
+		var/list/string_list = list()
+		for(var/datum/quirk/typepath in blacklist)
+			string_list += initial(typepath.name)
+		quirk_string_blacklist += string_list
```
Comment (spookydonut) [inline code/controllers/subsystem/processing/quirks.dm:51]: this doesn't work

Comment (LemonInTheDark) [reply code/controllers/subsystem/processing/quirks.dm:51]: I fucking

Comment (LemonInTheDark) [inline code/controllers/subsystem/processing/quirks.dm:51]:
```suggestion
		quirk_string_blacklist += list(string_list)
```
so they embed properly (the idea is entries in the same sublist conflict with each other and all)

Comment (spookydonut) [reply code/controllers/subsystem/processing/quirks.dm:51]: I LOVE BYOND

File: code/controllers/subsystem/processing/quirks.dm [hunk 2]
```diff
@@ -98,7 +109,7 @@ PROCESSING_SUBSYSTEM_DEF(quirks)
 	//Create a random list of stuff to start with
 	while(bonus_quirks > added_quirk_count)
 		var/quirk = pick(possible_quirks) //quirk is a string
-		if(quirk in quirk_blacklist) //prevent blacklisted
+		if(quirk in GLOB.quirk_blacklist) //prevent blacklisted
```
Comment (MrMelbert) [inline code/controllers/subsystem/processing/quirks.dm:112]: `quirk_blacklist` is a list of lists so this check does nothing

So this code has probably always been broken forever

File: code/controllers/subsystem/processing/quirks.dm [hunk 3]
```diff
@@ -113,7 +124,7 @@ PROCESSING_SUBSYSTEM_DEF(quirks)
 		if(!length(possible_quirks))//Lets not get stuck
 			break
 		var/quirk = pick(quirks)
-		if(quirk in quirk_blacklist) //prevent blacklisted
+		if(quirk in GLOB.quirk_blacklist) //prevent blacklisted
```
Comment (MrMelbert) [inline code/controllers/subsystem/processing/quirks.dm:127]: Same

File: code/controllers/subsystem/processing/quirks.dm [hunk 4]
```diff
@@ -128,7 +139,7 @@ PROCESSING_SUBSYSTEM_DEF(quirks)
...
```
Comment (MrMelbert) [inline code/controllers/subsystem/processing/quirks.dm:142]: Same

**Comment by MrMelbert** [comment]
Oh I just read the PR and we're intentionally leave it broken OK

File: code/controllers/subsystem/processing/quirks.dm [hunk 5]
```diff
@@ -1,6 +1,40 @@
...
+		list(/datum/quirk/prosthetic_organ, /datum/quirk/tin_man, /datum/quirk/body_purist),
+		list(/datum/quirk/quadruple_amputee, /datum/quirk/paraplegic, /datum/quirk/hemiplegic),
+		list(/datum/quirk/quadruple_amputee, /datum/quirk/frail),
+		list(/datum/quirk/social_anxiety, /datum/quirk/mute),
+		list(/datum/quirk/mute, /datum/quirk/softspoken),
+		list(/datum/quirk/poor_aim, /datum/quirk/bighands),
+		list(/datum/quirk/bilingual, /datum/quirk/foreigner),
+		list(/datum/quirk/spacer_born, /datum/quirk/paraplegic, /datum/quirk/item_quirk/settler),
+		list(/datum/quirk/photophobia, /datum/quirk/nyctophobia),
+		list(/datum/quirk/item_quirk/settler, /datum/quirk/freerunning)
```
Comment (san7890) [inline code/controllers/subsystem/processing/quirks.dm:25]:
```suggestion
		list(/datum/quirk/item_quirk/settler, /datum/quirk/freerunning),
```

**Comment by spookydonut** [comment]
> Oh I just read the PR and we're intentionally leave it broken OK

Yeah I've left that broken because i'm unclear on how that proc is meant to be using the blacklist, needless to say it has been broken a long time.

File: code/controllers/subsystem/processing/quirks.dm [hunk 6]
```diff
@@ -1,6 +1,40 @@
...
+		list(/datum/quirk/quadruple_amputee, /datum/quirk/paraplegic, /datum/quirk/hemiplegic),
+		list(/datum/quirk/quadruple_amputee, /datum/quirk/frail),
+		list(/datum/quirk/social_anxiety, /datum/quirk/mute),
+		list(/datum/quirk/mute, /datum/quirk/softspoken),
+		list(/datum/quirk/poor_aim, /datum/quirk/bighands),
+		list(/datum/quirk/bilingual, /datum/quirk/foreigner),
+		list(/datum/quirk/spacer_born, /datum/quirk/paraplegic, /datum/quirk/item_quirk/settler),
+		list(/datum/quirk/photophobia, /datum/quirk/nyctophobia),
+		list(/datum/quirk/item_quirk/settler, /datum/quirk/freerunning),
+	))
```
Comment (Mothblocks) [inline code/controllers/subsystem/processing/quirks.dm:26]: dedent by one

File: code/controllers/subsystem/processing/quirks.dm [hunk 7]
```diff
@@ -1,6 +1,40 @@
...
+
+GLOBAL_LIST_INIT(quirk_string_blacklist, generate_quirk_string_blacklist())
+
+/proc/generate_quirk_string_blacklist()
+	. = list()
+	for(var/blacklist in GLOB.quirk_blacklist)
+		var/list/string_list = list()
+		for(var/datum/quirk/typepath as anything in blacklist)
+			string_list += initial(typepath.name)
+		. += list(string_list)
```
Comment (Mothblocks) [inline code/controllers/subsystem/processing/quirks.dm:36]: avoid `.` for trivial cases, or `return .` at the end

**Pull request merged**